File: examples/BookShop.hs
{-# LANGUAGE TypeOperators
           , FlexibleContexts #-}
module BookShop where

import SessionCheck
import SessionCheck.Backend.Erlang

bookShop :: (Int :< t, String :< t, [Int] :< t) => [Int] -> Spec t ()
bookShop books = do 
  b <- send posInt --anyInt
  let books' = b : books

  choice <- choose ["another", "request"]
  case choice of
    "another" -> bookShop books'
    "request" -> request books'
            
request :: (Int :< t, String :< t, [Int] :< t) => [Int] -> Spec t ()
request books = do
  get $ permutationOf books

  choice <- choose ["another", "done"]
  case choice of
    "another" -> bookShop books
    "done"    -> stop

main :: IO ()
main = do
  checkCoherence (bookShop [])
  erlangMain "bookShop:main" (bookShop [])

File: examples/bookShop.erl
-module(bookShop).
-export([main/0]).

main() -> spawn(fun() -> f() end).

f() ->
  register(p, self()),
  receive 
    _ -> loop([])
  end.

loop(XS) ->
  receive 
    {_, I} -> continue(I, XS)
  end.

continue(I, XS) ->
  Lst = if
          %I < 0 -> XS;
          %length(XS) >= 5 -> XS;
          true  -> [I|XS]
        end,
  receive 
    {_,  "another"} -> loop(Lst);
    {Hs, "request"} -> Hs ! Lst, finish(Lst)
  end.

finish(XS) ->
  receive
    {_, "another"} -> loop(XS);
    {_, "done"}    -> exit(done)
  end.
